### ISSUE TITLE
TiffImageWriter: Unreadable LZW

### ISSUE THREAD

**Issue by Schmidor** [haraldk/TwelveMonkeys#257]
A read/write with the attached LZW compressed TIFF produces an unreadable image:

```
    @Test
    public void testReadWriteReadLZW() throws IOException{
        ImageInputStream input = ImageIO.createImageInputStream(getClass().getResource("/tiff/a33.tif"));
        ImageReader reader = ImageIO.getImageReaders(input).next();
        reader.setInput(input);

        BufferedImage original = reader.read(0);
        IIOMetadata metadata = reader.getImageMetadata(0);
        input.close();
        reader.dispose();

        ImageWriter writer = createImageWriter();
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        ImageOutputStream output = ImageIO.createImageOutputStream(buffer);
        writer.setOutput(output);
        writer.write(new IIOImage(original, null, metadata));
        output.close();
        writer.dispose();

        ImageInputStream input2 = ImageIO.createImageInputStream(new ByteArrayInputStream(buffer.toByteArray()));
        ImageReader reader2 = ImageIO.getImageReaders(input2).next();
        reader2.setInput(input2);
        reader2.read(0);
    }
```
```
java.io.EOFException: Unexpected end of stream
	at com.twelvemonkeys.imageio.plugins.tiff.HorizontalDeDifferencingStream.fetch(HorizontalDeDifferencingStream.java:95)
```

![a33](https://cloud.githubusercontent.com/assets/10960818/15674949/e0867930-273f-11e6-836d-10cb1887adf2.png)

**Comment by haraldk** [comment]
Thanks,

Could be related to the incorrect offsets bug, but it's computed differently for compressed... So, most likely there's a bug in the LZW encoder.

I'll look into it!

-- 
Harald K

**Comment by haraldk** [comment]
Hmm.. 

The writer doesn't seem to write the data correctly using other compressions either (testet none, ZLib and PackBits), due to a bug in how we write "packed" pixel formats (multiple pixels in one sample, in this case 1 bit, but using byte samples, so 8 pixels per byte)...

-- 
Harald K

**Comment by haraldk** [comment]
Arright. 

Found the bug. The problem is that the `LZWEncoder` needs to know the length of the data (it's a stupidity of the API that I haven't been able to fix yet). The length is computed incorrectly for < 8 bits, and because of this, the LZW EOI code is never written.

With that fixed, the image can be read without `EOFException`. :-)

Unfortunately, there are multiple other issues as well: 

* The `TIFFImageWriter` unconditionally adds Predictor to the encoding. Even though the original metadata specifies no predictor. I'm also unsure if predictor is allowed according to the spec, or commonly supported for B/W... At least Preview on OS X can't display it (it comes out all black). 
* If I remove the Predictor, it will be displayed, but the image data seems to be inverted (black is white and vice versa), as it is no longer normalized. So we need to take that into consideration as well.

Hope to have a fix soon!

-- 
Harald K

**Comment by haraldk** [comment]
That wasn't fun... But I fixed a lot of other bugs in the same go, so I guess I can live with it. 

Unfortunately, I think I broke your #239 PR in the mess. But I'll see if I can sort it out manually.

As usual, this will be released as 3.3, as soon as possible.


Best regards,

-- 
Harald K
